Annotate exceptions arriving over JSRPC as "remote"

jclee opened the pull request:
...so that they get the expected ".remote" property when tunneled to JS, like exceptions from Durable Object fetch() calls do.

jasnell commented:
Any possibility of having tests with this?

jclee commented:
> Any possibility of having tests with this?

- [ ] I've got some on the internal PR... Will try adding workerd-specific coverage.

jasnell reviewed (approved):
Lgtm assuming internal ci is good

kentonv commented:
A couple inconsistencies with the way "remote" is handled for fetch():

1. Here you've placed the logic on the client side to add the remote annotation, but for fetch() it happens server-side (in WorkerEntrypoint). Maybe we should keep these the same? I think this means putting the logic in `JsRpcTargetBase::call()` instead.
2. WorkerEntrypoint only adds the annotation to JSG exceptions, not to internal errors, but your code here doesn't seem to distinguish.

jclee commented:
> A couple inconsistencies with the way "remote" is handled for fetch():
> 
> 1. Here you've placed the logic on the client side to add the remote annotation, but for fetch() it happens server-side (in WorkerEntrypoint). Maybe we should keep these the same? I think this means putting the logic in `JsRpcTargetBase::call()` instead.
> 2. WorkerEntrypoint only adds the annotation to JSG exceptions, not to internal errors, but your code here doesn't seem to distinguish.

- [ ] Ah, makes sense... taking a look.

jclee commented:
I moved the `.remote` annotation from the client side of the RPC boundary to the server side, and added a check to ensure only tunneled (`jsg.`-prefixed) exceptions get the annotation.  (I wasn't sure how to test non-remote exceptions from within the wd-test framework, but I managed to add some coverage for it on the internal PR.)
